### ISSUE TITLE
default ignore files list

### ISSUE THREAD

**Issue by doronbehar** [ncw/rclone#2894]
#### Version:
```
rclone v1.45
- os/arch: linux/amd64
- go version: go1.11.2
```

#### Feature request

Well first of all, I'm surprised I haven't seen anyone suggesting this idea here or on the forum and I was surprised to see it wasn't implemented already. The idea is probably familiar to you from other programs:
Many sync programs like Syncthing and Git, ignore certain files when they encounter files like `.gitignore` and `.stignore` (Syncthing) which specify file names' patterns to ignore. It seems rclone has this functionality embedded in several optional flags but It could be great if it could by default search for a file with a default name such as `.rcloneignore` and act like Git or Syncthing with it.

**Comment by ncw** [comment]
I think what this needs is a pre-processing step to traverse the heirachy and look for `.gitignore` or `.rcloneignore` files and turn them into filters.

If a `.rcloneignore` file could have the same syntax as a `--filter-from` file that would make rclone's life easy.

I'm sure there is another issue about this somewhere but I can't find it!

Do you want to work on this @doronbehar  ?

**Comment by doronbehar** [comment]
Thanks for the offer, I have no experience at all with Go but I'll
consider diving into this through this issue.

**Comment by ncw** [comment]
> Thanks for the offer, I have no experience at all with Go but I'll consider diving into this through this issue.

That would be great!  Let me know if you need help.

**Comment by doronbehar** [comment]
Hey, It's been a while..

I've gained some basic experience in Go recently so I decided to try dive into this idea. The code-base is pretty complex! For a start, I couldn't even find where's the iteratation over every file in a source / destination. There's the function [`syncCopyMove.run`](https://github.com/ncw/rclone/blob/fc4f20d52ff465a1e1eb0bbf43668310ac3070a9/fs/sync/sync.go#L636) which should be responsible of that but it seems to be wrapping a few `syncCopyMove.start*` and `syncCopyMove.stop*` functions. I've also noticed it runs [`March.Run()`](https://github.com/ncw/rclone/blob/fc4f20d52ff465a1e1eb0bbf43668310ac3070a9/fs/march/march.go#L112) as well. As I go deeper I see less and less relation to actual file read operations. Anyway, I'm kind of lost..

Would it be possible to work this out without touching those functions? I was thinking perhaps to patch [`fs/filter/filter.go`](https://github.com/ncw/rclone/blob/fc4f20d52ff465a1e1eb0bbf43668310ac3070a9/fs/filter/filter.go#L116-L117). What do you think?

**Comment by ncw** [comment]
> Hey, It's been a while..
> 
> I've gained some basic experience in Go recently so I decided to try dive into this idea. 

Great!

> The code-base is pretty complex!

Yes indeed.  It is complicated by the fact that rclone runs lots of thing in parallel so communication is done over channels.

> For a start, I couldn't even find where's the iteratation over every file in a source / destination. There's the function [`syncCopyMove.run`](https://github.com/ncw/rclone/blob/fc4f20d52ff465a1e1eb0bbf43668310ac3070a9/fs/sync/sync.go#L636) which should be responsible of that but it seems to be wrapping a few `syncCopyMove.start*` and `syncCopyMove.stop*` functions. I've also noticed it runs [`March.Run()`](https://github.com/ncw/rclone/blob/fc4f20d52ff465a1e1eb0bbf43668310ac3070a9/fs/march/march.go#L112) as well. As I go deeper I see less and less relation to actual file read operations. Anyway, I'm kind of lost..

march is where the actual iteration over a source and a destination file system occurs.

> Would it be possible to work this out without touching those functions? I was thinking perhaps to patch [`fs/filter/filter.go`](https://github.com/ncw/rclone/blob/fc4f20d52ff465a1e1eb0bbf43668310ac3070a9/fs/filter/filter.go#L116-L117). What do you think?

I think what we need to do is load up these `.rcloneignore` files when they are found in the source heirachy.

For a sync/copy/move just before here would probably be the right place

https://github.com/ncw/rclone/blob/fc4f20d52ff465a1e1eb0bbf43668310ac3070a9/fs/march/march.go#L387

Though the listing has already been filtered there. and if we want `.rcloneignore` files to be able to filter then directory they are in then we'll need to either re-filter the source, or load the `.rcloneignore` file a bit earlier....

I see what you mean, loading the `.rcloneignore` files as we go a long is quite tricky!

What about loading them at the start, so do a traversal of the source first?  If we are doing that then it would need to be gated on a flag.  What do you think about that - should this feature have a flag to enable it?  Say `--filter-from-file .rcloneignore`?

**Comment by doronbehar** [comment]
Thanks for giving these pointers.. Though the intuitive place to read those `.rcloneignore` files is in [`fs/march/march.go`](https://github.com/ncw/rclone/blob/fc4f20d52ff465a1e1eb0bbf43668310ac3070a9/fs/march/march.go#L387), I still tend to think it'll be better to hack this into [the `NewFilter` function in `fs/filter/filter.go`](https://github.com/ncw/rclone/blob/fc4f20d52ff465a1e1eb0bbf43668310ac3070a9/fs/filter/filter.go#L116-L118).

What if when it's `opt` is `nil`, we'll recursively read all `.rcloneignore` files in the source directory (not sure yet how to access it from there) and create a [`Filter`](https://github.com/ncw/rclone/blob/fc4f20d52ff465a1e1eb0bbf43668310ac3070a9/fs/filter/filter.go#L105-L106) according to all of them?

**Comment by 131** [comment]
Hi, could anyone get a look into this ?

**Comment by Emasoft** [comment]
We really need this.

**Comment by Valemos** [comment]
This feature would be really useful to manage large file trees, where each directory is isolated and has it's own stuff to be ignored inside it.

**Comment by fjih** [comment]
I took a loose look at the code to see how difficult this would be. I'm fairly new to golang, but it looks to me that this may be more difficult that initially thought due to the source/destination overlap check. The overlap check does handle the case where the overlap is filtered out via the filter flags, but that only works as the filters are loaded before the check is performed.

In `sync` the overlap check is performed before we get to `march`, so enumeration of the source looking for the ignore files would have to be its own separate read before the overlap check occurs.

I suppose alternatively the ignore files could only apply to `copy`. The overlap check could also just not support being filtered by the ignore files.
